native/dlaqr4.go

```diff
@@ -0,0 +1,467 @@
+		case ilo > 0 && h[ilo*ldh+ilo-1] != 0:
+			panic("lapack: block not isolated")
+		case ihi+1 < n && h[(ihi+1)*ldh+ihi] != 0:
+			panic("lapack: block not isolated")
+		case len(wr) != ihi+1:
+			panic("lapack: bad length of wr")
+		case len(wi) != ihi+1:
+			panic("lapack: bad length of wi")
+		}
+	}
+
+	// Quick return for n=0.
+	if n == 0 {
+		work[0] = 1
+		return 0
+	}
+
+	if n <= ntiny {
+		// Tiny matrices must use Dlahqr.
+		if lwork != -1 {
+			unconverged = impl.Dlahqr(wantt, wantz, n, ilo, ihi, h, ldh, wr, wi,
+				iloz, ihiz, z, ldz)
+		}
+		work[0] = 1
+		return unconverged
+	}
+
+	// Use small bulge multi-shift QR with aggressive early deflation on
+	// larger-than-tiny matrices.
+	var jbcmpz string
+	if wantt {
+		jbcmpz = "S"
+	} else {
+		jbcmpz = "E"
+	}
+	if wantz {
+		jbcmpz += "V"
+	} else {
+		jbcmpz += "N"
+	}
+
+	// nwr is the recommended deflation window size. n is greater than 11,
+	// so there is enough subdiagonal workspace for nwr >= 2 as required.
+	// (In fact, there is enough subdiagonal space for nwr >= 3.)
+	// TODO(vladimir-ch): If there is enough space for nwr >= 3, should we
+	// use it?
+	nwr := impl.Ilaenv(13, "DLAQR4", jbcmpz, n, ilo, ihi, lwork)
+	nwr = max(2, nwr)
+	nwr = min(ihi-ilo+1, min((n-1)/3, nwr))
+
+	// nsr is the recommended number of simultaneous shifts. n is greater
+	// than 11, so there is enough subdiagonal workspace for nsr to be even
+	// and greater than or equal to two as required.
+	nsr := impl.Ilaenv(15, "DLAQR4", jbcmpz, n, ilo, ihi, lwork)
+	nsr = min(nsr, min((n+6)/9, ihi-ilo))
+	nsr = max(2, nsr-nsr%2)
+
+	// Workspace query call to Dlaqr2.
+	impl.Dlaqr2(wantt, wantz, n, ilo, ihi, nwr+1, nil, 0, iloz, ihiz, nil, 0,
+		nil, nil, nil, 0, n, nil, 0, n, nil, 0, work, -1)
+	// Optimal workspace is max(Dlaqr5, Dlaqr2).
+	lwkopt := max(3*nsr/2, int(work[0]))
+	// Quick return in case of workspace query.
+	if lwork == -1 {
+		work[0] = float64(lwkopt)
+		return 0
+	}
+
+	// Dlahqr/Dlaqr4 crossover point.
+	nmin := impl.Ilaenv(12, "DLAQR4", jbcmpz, n, ilo, ihi, lwork)
+	nmin = max(ntiny, nmin)
+
+	// Nibble determines when to skip a multi-shift QR sweep (Dlaqr5).
+	nibble := impl.Ilaenv(14, "DLAQR4", jbcmpz, n, ilo, ihi, lwork)
+	nibble = max(0, nibble)
+
+	// Computation mode of far-from-diagonal orthogonal updates in Dlaqr5.
+	kacc22 := impl.Ilaenv(16, "DLAQR4", jbcmpz, n, ilo, ihi, lwork)
+	kacc22 = max(0, min(kacc22, 2))
+
+	// nwmax is the largest possible deflation window for which there is
+	// sufficient workspace.
+	nwmax := min((n-1)/3, lwork/2)
+
+	// nsmax is the largest number of simultaneous shifts for which there is
+	// sufficient workspace.
+	nsmax := min((n+6)/9, 2*lwork/3)
+	nsmax -= nsmax % 2
+
+	var (
+		ndec int     // Deflation window size decrement.
+		ndfl = 1     // Number of iteration since last deflation.
+		nw   = nwmax // Start with maximum deflation window size.
+
+		itmax = max(30, 2*kexsh) * max(10, (ihi-ilo+1))
+		it    = 0
+	)
+	// Main loop.
```

Comment:
Blank line above, or if the var decl block should be considered part of this, put the comment above that.